File: src/main/java/zielu/gittoolbox/completion/CompletionServiceImpl.java
package zielu.gittoolbox.completion;

import com.google.common.collect.ImmutableList;
import com.intellij.openapi.Disposable;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.project.Project;
import git4idea.repo.GitRepository;
import java.io.File;
import java.lang.ref.WeakReference;
import java.util.Collection;
import java.util.List;
import org.jetbrains.annotations.NotNull;
import zielu.gittoolbox.config.GitToolBoxConfigPrj;
import zielu.gittoolbox.formatter.Formatter;

class CompletionServiceImpl implements CompletionService, Disposable {
  private final Logger log = Logger.getInstance(getClass());
  private final CompletionFacade facade;
  private WeakReference<CompletionScopeProvider> scopeProviderRef;
  private volatile ImmutableList<Formatter> formatters;

  CompletionServiceImpl(@NotNull Project project) {
    facade = new CompletionFacade(project);
  }

  @Override
  public void onConfigChanged(@NotNull GitToolBoxConfigPrj config) {
    synchronized (this) {
      formatters = null;
    }
  }

  @Override
  public void setScopeProvider(@NotNull CompletionScopeProvider scopeProvider) {
    log.debug("Set scope provider: ", scopeProvider);
    scopeProviderRef = new WeakReference<>(scopeProvider);
  }

  @Override
  @NotNull
  public Collection<GitRepository> getAffected() {
    CompletionScopeProvider scopeProvider = getScopeProvider();
    Collection<File> affectedFiles = scopeProvider.getAffectedFiles();
    log.debug("Get affected files: ", affectedFiles);
    Collection<GitRepository> affectedRepositories = findAffectedRepositories(affectedFiles);
    log.debug("Get affected repositories: ", affectedRepositories);
    return affectedRepositories;
  }

  private CompletionScopeProvider getScopeProvider() {
    if (scopeProviderRef != null) {
      CompletionScopeProvider provider = scopeProviderRef.get();
      if (provider != null) {
        return provider;
      }
    }
    return CompletionScopeProvider.EMPTY;
  }

  private Collection<GitRepository> findAffectedRepositories(Collection<File> affectedFiles) {
    return facade.getRepositories(affectedFiles);
  }

  @Override
  @NotNull
  public List<Formatter> getFormatters() {
    if (formatters == null) {
      synchronized (this) {
        if (formatters == null) {
          formatters = facade.getFormatters();
        }
      }
    }
    return formatters;
  }

  @Override
  public void dispose() {
    formatters = null;
    scopeProviderRef = null;
  }
}

File: src/main/kotlin/zielu/gittoolbox/completion/CompletionServiceImpl.kt
package zielu.gittoolbox.completion

import com.intellij.openapi.Disposable
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.Project
import git4idea.repo.GitRepository
import zielu.gittoolbox.formatter.Formatter
import java.io.File
import java.lang.ref.WeakReference

internal class CompletionServiceImpl(
  project: Project
) : CompletionService, Disposable {
  private val facade: CompletionFacade = CompletionFacade(project)
  private var scopeProviderRef: WeakReference<CompletionScopeProvider>? = null
  @Volatile
  private var formatters: List<Formatter>? = null

  override fun setScopeProvider(scopeProvider: CompletionScopeProvider) {
    log.debug("Set scope provider: ", scopeProvider)
    scopeProviderRef = WeakReference(scopeProvider)
    clearFormatters()
  }

  private fun clearFormatters() {
    synchronized(this) {
      formatters = null
      log.info("Clear formatters")
    }
  }

  override fun getAffected(): Collection<GitRepository> {
    val scopeProvider = getScopeProvider()
    val affectedFiles = scopeProvider.getAffectedFiles()
    log.debug("Get affected files: ", affectedFiles)
    val affectedRepositories = findAffectedRepositories(affectedFiles)
    log.debug("Get affected repositories: ", affectedRepositories)
    return affectedRepositories
  }

  private fun getScopeProvider(): CompletionScopeProvider {
    return scopeProviderRef?.get() ?: CompletionScopeProvider.empty
  }

  private fun findAffectedRepositories(affectedFiles: Collection<File>): Collection<GitRepository> {
    return facade.getRepositories(affectedFiles)
  }

  override fun getFormatters(): List<Formatter> {
    if (formatters == null) {
      synchronized(this) {
        if (formatters == null) {
          formatters = facade.getFormatters()
          log.debug("Current formatters: ", formatters)
        }
      }
    }
    return formatters!!
  }

  override fun dispose() {
    clearFormatters()
    scopeProviderRef = null
  }

  private companion object {
    private val log = Logger.getInstance(CompletionServiceImpl::class.java)
  }
}
